src/workspace-resolver.ts

```diff
@@ -0,0 +1,152 @@
+import path from 'node:path'
+import yaml from 'js-yaml'
+import { readFileContents } from './fs-utils'
+import { IGlobber } from './globber'
+import { ensureWorkspacesArray, uniq } from './util'
+import type { IPackageFile } from './types'
+
+/**
+ * Files service.
+ */
+export interface IWorkspaceResolverService {
+  /**
+   * Reads, parses, and normalizes a workspaces configuration from the following files, in this order:
+   * - `package.json` `workspaces` field, as an array of globs.
+   * - `package.json` `workspaces` field, as an object with a `projects` field, which is an array of globs.
+   * - `pnpm-workspace.yaml` `packages` field, as an array of globs.
+   *
+   * Path is relative to the current working directory.
+   */
+  getWorkspaces(root: string, globber: IGlobber): Promise<IWorkspacesArray>
+}
+
+/**
+ * @example
+ * ```json
+ * "workspaces": [
+ *  "packages/*",
+ * ]
+ * ```
+ */
+export type IWorkspacesArray = Array<string>
+
+/**
+ * @example
+ * ```yaml
+ * projects:
+ * - 'packages/*'
+ * ```
+ */
+export interface IWorkspacesObject {
+  packages: IWorkspacesArray
+}
+
+/**
+ * @see {@link IWorkspacesArray}
+ */
+type NpmWorkspacesConfig = IWorkspacesArray
+
+/**
+ * Yarn is a special snowflake.
+ *
+ * @example
+ * ```json
+ * "workspaces": {
+ *   "packages": [
+ *     "packages/*",
+ *   ],
+ *   "nohoist": []
+ * }
+ * ```
+ */
+type YarnWorkspacesConfig =
+  | IWorkspacesArray
+  | (IWorkspacesObject & { nohoist?: Array<string> })
+
+/**
+ * The contents of a `pnpm-workspace.yaml` file.
+ *
+ * @example
+ * ```yaml
+ * packages:
+ * - 'packages/*'
+ * ```
+ */
+export type PnpmWorkspacesConfig = IWorkspacesObject
+
+/**
+ * @see {@link IWorkspacesArray}
+ */
+type BunWorkspacesConfig = IWorkspacesArray
+
+/**
+ * Section in `package.json` representing workspaces.
+ */
+export type IWorkspacesSection =
+  | NpmWorkspacesConfig
+  | YarnWorkspacesConfig
+  | BunWorkspacesConfig
+
+export function createWorkspaceResolverService(): IWorkspaceResolverService {
+  return {
+    getWorkspaces: async (root, globber) =>
+      await getGlobbedWorkspaces(root, globber),
+  }
+}
+
+async function getGlobbedWorkspaces(root: string, globber: IGlobber) {
+  const workspaces = await getWorkspacesArray(root)
+
+  const manifests = await Promise.all(
+    workspaces.map(
+      async (workspace) => await globber.globPackageFiles(workspace),
+    ),
+  )
+
+  return uniq(manifests.flat())
+}
+
+async function getWorkspacesArray(root: string): Promise<IWorkspacesArray> {
```

Comment:
I think these 2 functions can be consolidated into one, and they could probably be moved into the `getWorkspaces` of the service implementation.

Reply:
Ok!